sup3r/preprocessing/base.py

```diff
@@ -0,0 +1,414 @@
+"""Base container classes - object that contains data. All objects that
+interact with data are containers. e.g. loaders, extracters, data handlers,
+samplers, batch queues, batch handlers.
+"""
+
+import logging
+import pprint
+from abc import ABCMeta
+from collections import namedtuple
+from typing import ClassVar, Dict, Optional, Tuple, Union
+from warnings import warn
+
+import numpy as np
+import xarray as xr
+
+import sup3r.preprocessing.accessor  # noqa: F401 # pylint: disable=W0611
+from sup3r.preprocessing.accessor import Sup3rX
+from sup3r.preprocessing.utilities import (
+    _log_args,
+    get_composite_signature,
+    get_source_type,
+)
+from sup3r.typing import T_Dataset
+
+logger = logging.getLogger(__name__)
+
+
+class Sup3rDataset:
+    """Interface for interacting with one or two `xr.Dataset` instances
+    This is either a simple passthrough for a `xr.Dataset` instance or a
+    wrapper around two of them so they work well with Dual objects like
+    DualSampler, DualExtracter, DualBatchHandler, etc...)
+
+    Note
+    ----
+    (1) This may seem similar to :class:`Collection`, which also can
+    contain multiple data members, but members of :class:`Collection` objects
+    are completely independent while here there are at most two members which
+    are related as low / high res versions of the same underlying data.
+
+    (2) Here we make an important choice to use high_res members to compute
+    means / stds. It would be reasonable to instead use the average of high_res
+    and low_res means / stds for aggregate stats but we want to preserve the
+    relationship between coarsened variables after normalization (e.g.
+    temperature_2m, temperature_max_2m, temperature_min_2m). This means all
+    these variables should have the same means and stds, which ultimately come
+    from the high_res non coarsened variable.
+    """
+
+    def __init__(
+        self,
+        data: Optional[Union[tuple, T_Dataset]] = None,
+        **dsets: Union[xr.Dataset, Sup3rX],
+    ):
```

Comment:
Yeah Ill add some stuff here. Im confused on the naming recs though? Would data_hr be a singlet and data_hr_lr a doublet?

Reply:
Yeah you basically have handling for multiple options for data, right? if data is tuple do this... if data is a single do this... I'm saying you should explicitly provide kwargs data_hr which would be a single interpreted as the high-res data or data_hr_lr/data_pair which would be the high-res/low-res tuple. This makes it more explicit what you expect and what the priorities are... I think we should prefer explicit kwarg options rather than describing the options in the docstring for a single arg. 

Separate note, in this docstring you say dsets is the prefered option as a dictionary... You should say it's flexible kwargs and then name the keys. The example in the class doc really explains this better than the dsets docstring. 

Last note: i don't think the word "singleton" should be used in this case. This has a specific definition in design patterns that doesn't match what you're describing.

Reply:
Yeah good call on the terminology. The reason I didnt make it explicit is because the `.data` member of all container classes gets wrapped as a `Sup3rDataset` through `Container@setter.data`. This includes containers with a single data member and dual containers. I'd have to put these checks in multiple places if they were removed from this init.